src/disco/shred/fd_fec_resolver.c

```diff
@@ -0,0 +1,527 @@
+  fd_tcache_t  * done          = resolver->done;
+  ulong          done_depth    = fd_tcache_depth       ( done );
+  ulong          done_map_cnt  = fd_tcache_map_cnt     ( done );
+  ulong        * done_oldest   = fd_tcache_oldest_laddr( done );
+  ulong        * done_ring     = fd_tcache_ring_laddr  ( done );
+  ulong        * done_map      = fd_tcache_map_laddr   ( done );
+
+  ulong partial_depth = resolver->partial_depth;
+
+  fd_fec_set_t * * free_list        = resolver->free_list;
+  fd_fec_set_t * * complete_list    = resolver->complete_list;
+  void         * * bmtree_free_list = resolver->bmtree_free_list;
+  set_ctx_t    *   curr_map         = resolver->curr_map;
+
+  fd_reedsol_t * reedsol       = resolver->reedsol;
+  fd_sha512_t  * sha512        = resolver->sha512;
+
+  /* Invariants:
+      * no key is in both the done tcache and the current map
+      * each set pointer provided to the new function is in exactly one
+          of curr_map, freelist, or complete_list
+      * bmtree_free_list has exactly partial_depth fewer elements than
+          freelist
+   */
+
+  int dup = 0;
+  (void)dup;
+  /* Note: we identify FEC sets by the first 64 bits of their signature.
+     Given how slow Ed25519 is and how short of a time these are
+     relevant for, this seems safe, but we should research the issue
+     further. */
+
+  /* Are we already done with this FEC set? */
+  /* TODO: xor this with something semi-secret to prevent someone from
+     spending a lot of compute to make a shred with a 0 signature which
+     would be interpreted as an invalid key. */
+  ulong signature = fd_ulong_load_8( shred->signature );
+  int found;
+  ulong map_idx;
+
+  (void)map_idx;
+  FD_TCACHE_QUERY( found, map_idx, done_map, done_map_cnt, signature );
+  if( found )  return FD_FEC_RESOLVER_SHRED_IGNORED; /* With no packet loss, we expect found==1 about 50% of the time */
+
+  set_ctx_t * ctx = ctx_map_query( curr_map, signature, NULL );
+
+  fd_bmtree_node_t _leaf[1];
+  uchar variant    = shred->variant;
+  uchar shred_type = fd_shred_type( variant );
+
+  ulong tree_depth           = fd_shred_merkle_cnt( variant );
+  ulong reedsol_protected_sz = 1115UL - 20UL*tree_depth + 0x58UL - 0x40UL;
+  ulong merkle_protected_sz  = reedsol_protected_sz + fd_ulong_if( shred_type==FD_SHRED_TYPE_MERKLE_DATA, 0UL, 0x59UL - 0x40UL );
+
+  fd_bmtree_hash_leaf( _leaf, (uchar const *)shred + sizeof(fd_ed25519_sig_t), merkle_protected_sz, FD_BMTREE_LONG_PREFIX_SZ );
+
+  /* code_idx: if this is a Merkle coding shred, the index of it in the
+     Merkle array, otherwise 0. */
+  ulong code_idx   = fd_ulong_if( shred_type==FD_SHRED_TYPE_MERKLE_DATA, 0UL, shred->code.idx );
+  ulong shred_idx  = fd_ulong_if( shred_type==FD_SHRED_TYPE_MERKLE_DATA,
+                                  shred->idx - shred->fec_set_idx,
+                                  shred->code.data_cnt + code_idx );
+  /* TODO: bounds check shred_pos. make sure proof depth has the right
```

Comment:
Bounds checks again... this is unverified input right? We need to make this bulletproof.

Reply:
Thanks, I rejiggered the indices and added the appropriate checks.